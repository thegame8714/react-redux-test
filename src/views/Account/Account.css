.account {
    border: 1px solid #000;
    width: 50%;
    margin-left: auto;
    margin-right: auto;
    margin-top: 20px;
    padding: 10px;
    text-align: left;
}

.account .account-item {
    margin: 10px 0;
}

.account .account-item .label {
    color: grey;
    font-size: 12px;
    width: 100%;
}

.account .left-column {
    width: 50%;
}

.account-actions {
    float: right;
}

.account-actions button {
    margin-left: 20px;
    

}

.button {
    color: #fff;
    background-color: #000;
    border: none;
    padding: 5px;
}

@media (max-width: 1044px) {
    .account-actions {
        float: none;
    }
    .account-actions button {
        margin-right: 20px;
        margin-left: 0
    }

}

@media (max-width: 768px) {
    .account-actions button {
        width: 100%;
        margin: 5px 0;
    }

}