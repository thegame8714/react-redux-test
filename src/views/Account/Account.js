import React from 'react'
import { connect } from 'react-redux'
import { action as accountActions } from '../../redux/modules/account/actionCreators'
import AccountActions from '../../components/AccountActions/AccountActions'
import './Account.css'

export class Account extends React.Component {

    static propTypes = {
        retrieveAccountInfo: React.PropTypes.func,
        topUpAccount: React.PropTypes.func,
        withdrawAccount: React.PropTypes.func,
        account: React.PropTypes.object,
        isLoadingAccount: React.PropTypes.bool
    }

    componentDidMount() {
        this.props.retrieveAccountInfo()
    }

    updateAccountStatus() {
        this.props.toggleAccountStatus()
    }

    topUpAccount() {
        this.props.topUpAccount()
    }

    withdrawAccount() {
        this.props.withdrawAccount()
    }

    render() {
        const {
            account,
            isLoadingAccount 
        } = this.props

        return (
            <div className="account">
                {isLoadingAccount ? <div className="loader">Account loading....</div> :
                <span className="left-column">
                    <div className="account-item">
                        <div className="label">Account holder:</div> {account.holder.first_name} {account.holder.last_name}
                    </div>
                    <div className="account-item">
                        <div className="label">Amount available in the account:</div> {account.amount.currency === "USD" ? "$" : ""}{account.amount.amount}
                        {account.active && 
                        <AccountActions 
                            amount={account.amount.amount} 
                            maxAmount="10000000"
                            topUpAccount={this.topUpAccount.bind(this)}
                            withdrawAccount={this.withdrawAccount.bind(this)} />
                        }
                    </div>
                    <div className="account-item">
                        <div className="label">Account status:</div> {account.active ? <span className="active">ENABLED</span> : <span className="inactive">DISABLED</span>}</div>
                    <button className="toggleStatus button" onClick={this.updateAccountStatus.bind(this)}>{account.active ? "DISABLE ACCOUNT" : "ENABLE ACCOUNT"}</button>
                </span>
            }            
            </div>
        )
    }
} 

function mapStateToProps(state) {
    return {
        account: state.account.account,
        isLoadingAccount: state.account.isLoadingAccount
    }
}

export default connect(mapStateToProps, {
    retrieveAccountInfo: accountActions.retrieveAccountInfo,
    toggleAccountStatus: accountActions.toggleAccountStatus,
    topUpAccount: accountActions.topUpAccount,
    withdrawAccount: accountActions.withdrawAccount
})(Account)