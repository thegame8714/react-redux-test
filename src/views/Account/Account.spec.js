import React from 'react'
import { Account } from './Account'
import { shallow } from 'enzyme'
import chai from 'chai'
import sinon from 'sinon'
import renderer from 'react-test-renderer'

describe('Account', () => {
    let componentDidMountStub

    beforeEach(() => {
        componentDidMountStub = sinon.stub(Account.prototype, 'componentDidMount')
    })

    afterEach(() => {
        componentDidMountStub.restore()
    })

    it('should call the topUp function ', () => {
        const account = {
            id: "000001",
            holder: {
                first_name: "Fabio",
                last_name: "Salimbeni"
            },
            amount: {
                amount: "1000000",
                currency: "USD"
            },
            active: true
        }

        const topUpAccount = sinon.spy()

        const accountComponent = shallow(<Account retrieveAccountInfo={sinon.spy()}
                                                                account={account}
                                                                topUpAccount={topUpAccount} 
                                                                withdrawAccount={sinon.spy()}
                                                                isLoadingAccount={true} />)

        const inst = accountComponent.instance()
        inst.topUpAccount()
        chai.expect(topUpAccount).to.be.called
    })

     it('should call the withdraw function', () => {
        const account = {
            id: "000001",
            holder: {
                first_name: "Fabio",
                last_name: "Salimbeni"
            },
            amount: {
                amount: "1000000",
                currency: "USD"
            },
            active: true
        }

        const withdrawAccount = sinon.spy()

        const accountComponent = shallow(<Account retrieveAccountInfo={sinon.spy()}
                                                                account={account}
                                                                topUpAccount={sinon.spy()} 
                                                                withdrawAccount={withdrawAccount}
                                                                isLoadingAccount={true} />)

        const inst = accountComponent.instance()
        inst.withdrawAccount()
        chai.expect(withdrawAccount).to.be.called
    })


     it('should call the withdraw function', () => {
        const account = {
            id: "000001",
            holder: {
                first_name: "Fabio",
                last_name: "Salimbeni"
            },
            amount: {
                amount: "1000000",
                currency: "USD"
            },
            active: true
        }

        const updateAccountStatus = sinon.spy()

        const accountComponent = shallow(<Account retrieveAccountInfo={sinon.spy()}
                                                                account={account}
                                                                topUpAccount={sinon.spy()} 
                                                                withdrawAccount={sinon.spy()}
                                                                isLoadingAccount={false} 
                                                                toggleAccountStatus={sinon.spy()} />)

        accountComponent.find('.toggleStatus').simulate('click')
        chai.expect(updateAccountStatus).to.be.called
    })

     it('should render the page while the account is loading', () => {
         const account = {
            id: "000001",
            holder: {
                first_name: "Fabio",
                last_name: "Salimbeni"
            },
            amount: {
                amount: "1000000",
                currency: "USD"
            },
            active: true
        }
        const accountComponent = renderer.create(<Account retrieveAccountInfo={sinon.spy()}
                                                                account={account}
                                                                topUpAccount={sinon.spy()} 
                                                                withdrawAccount={sinon.spy()}
                                                                isLoadingAccount={true} />)
        
        expect(accountComponent.toJSON()).toMatchSnapshot()
    })

    it('should render the page after the account is loaded', () => {
        const account = {
            id: "000001",
            holder: {
                first_name: "Fabio",
                last_name: "Salimbeni"
            },
            amount: {
                amount: "1000000",
                currency: "USD"
            },
            active: true
        }
        const accountComponent = renderer.create(<Account retrieveAccountInfo={sinon.spy()}
                                                                account={account}
                                                                topUpAccount={sinon.spy()} 
                                                                withdrawAccount={sinon.spy()}
                                                                isLoadingAccount={false} />)
        
        expect(accountComponent.toJSON()).toMatchSnapshot()
    })
})