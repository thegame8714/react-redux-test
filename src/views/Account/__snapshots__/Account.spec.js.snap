exports[`Account should render the page after the account is loaded 1`] = `
<div
  className="account">
  <span
    className="left-column">
    <div
      className="account-item">
      <div
        className="label">
        Account holder:
      </div>
       
      Fabio
       
      Salimbeni
    </div>
    <div
      className="account-item">
      <div
        className="label">
        Amount available in the account:
      </div>
       
      $
      1000000
      <div
        className="account-actions">
        <button
          className="topUp button"
          disabled=""
          onClick={[Function]}>
          TOP UP $1000 ON ACCOUNT
        </button>
        <button
          className="withdraw button"
          disabled=""
          onClick={[Function]}>
          WITHDRAW $1000 FROM ACCOUNT
        </button>
      </div>
    </div>
    <div
      className="account-item">
      <div
        className="label">
        Account status:
      </div>
       
      <span
        className="active">
        ENABLED
      </span>
    </div>
    <button
      className="toggleStatus button"
      onClick={[Function]}>
      DISABLE ACCOUNT
    </button>
  </span>
</div>
`;

exports[`Account should render the page while the account is loading 1`] = `
<div
  className="account">
  <div
    className="loader">
    Account loading....
  </div>
</div>
`;
